### PR TITLE
[Role] az ad app permission delete: Support `--api-permissions` to delete specific `ResourceAccess`

### LINKED ISSUE #11106: az ad app permission delete - remove a particular resourceAccess permission
I want to be able to remove an api permission without having to remove the entire api and re-add api permissions.

I am writing a (powershell) script which compares the existing api's and api permissions on an application with a provided manifest.  The goal is to have the azure application updated with the values provided on the manifest.

In the event that an api permission needs to be deleted from the azure application there is currently no way to specifiy which api permision to delete, only the entire api can be removed.

For example if I have an azureAd application with the following required resource access

```
		{
			"resourceAppId": "2d4d3d8e-2be3-4bef-9f87-7875a61c29de",
			"resourceAccess": [
				{
					"id": "27e50e9e-ca8b-4cc6-8dd0-4ba2d49d8228",
					"type": "Role"
				},
				{
					"id": "10af711e-5051-4838-8ae7-9767c43d8a9c",
					"type": "Role"
				}
			]
		}
```

and I want to remove one of the permissions specified, I cannot only remove the required permission.  Instead I have to remove the entire api access and then add a new api permission using

```
az ad app permission add --api -api-permissions -id
```

I want to have a command 

```
az ad app permission delete --api $resouceAppId --id $applicationId --api-permissions-Id $resourceAccess.Id 
```

that removes only a particular apiPermssion.

**Describe alternatives you've considered**
removing the api access  and re-adding the still rquired permissions.

### PR DESCRIPTION
**Pull request by jiasli**
## Description

Resolve #11106: `az ad app permission delete`: Support `--api-permissions` to delete specific `ResourceAccess`

In `az ad app permission add`, both `--api` (`RequiredResourceAccess`) and `--api-permissions` (`ResourceAccess`) should be specified. 

But in `az ad app permission delete`, only `--api` can be specified. This PR adds the `--api-permissions` option to delete specific `ResourceAccess`.

The help message is also refined according to [the REST spec](https://github.com/Azure/azure-rest-api-specs/blob/32e56f061668a1bf1eeca6209000fad4c09afca8/specification/graphrbac/data-plane/Microsoft.GraphRbac/stable/1.6/graphrbac.json#L2817).

Before:

```
> az ad app permission add -h
...
Arguments
    --api             [Required] : The target API to access.
    --api-permissions [Required] : Space separated list of `<resource-access-id>=<type>`.

> az ad app permission delete -h
...
Arguments
    --api   [Required] : The target API to access.
```

After:

```
> az ad app permission add -h
...
Arguments
    --api             [Required] : Specify `RequiredResourceAccess.resourceAppId` - The unique
                                   identifier for the resource that the application requires access
                                   to. This should be equal to the appId declared on the target
                                   resource application.
    --api-permissions [Required] : Specify `ResourceAccess.id` - The unique identifier for one of
                                   the OAuth2Permission or AppRole instances that the resource
                                   application exposes. Space-separated list of `<resource-access-
                                   id>=<type>`.

> az ad app permission delete -h
...
Arguments
    --api   [Required] : Specify `RequiredResourceAccess.resourceAppId` - The unique identifier for
                         the resource that the application requires access to. This should be equal
                         to the appId declared on the target resource application.
    --api-permissions  : Specify `ResourceAccess.id` - The unique identifier for one of the
                         OAuth2Permission or AppRole instances that the resource application
                         exposes.
```

## Testing Guide

```sh
# Create application
az ad sp create-for-rbac --skip-assignment
$appId = "b9df9246-f8c2-4505-8556-814b7ecc95a9"

# Add MS Graph permissions
# 00000003-0000-0000-c000-000000000000: MS Graph resource ID
# 0e263e50-5827-48a4-b97c-d940288653c7: Delegated Directory.AccessAsUser.All
# e1fe6dd8-ba31-4d61-89e7-88639da4683d: Delegated User.Read
az ad app permission add --id $appId --api 00000003-0000-0000-c000-000000000000 --api-permissions 0e263e50-5827-48a4-b97c-d940288653c7=Scope e1fe6dd8-ba31-4d61-89e7-88639da4683d=Scope

# Add AD Graph permissions
# 00000002-0000-0000-c000-000000000000: AD Graph resource ID
# a42657d6-7f20-40e3-b6f0-cee03008a62a: Delegated Directory.AccessAsUser.All
# 311a71cc-e848-46a1-bdf8-97ff7156d8e6: Delegated User.Read
az ad app permission add --id $appId --api 00000002-0000-0000-c000-000000000000 --api-permissions a42657d6-7f20-40e3-b6f0-cee03008a62a=Scope 311a71cc-e848-46a1-bdf8-97ff7156d8e6=Scope

# List permissions
az ad app permission list --id $appId

# Delete whole --api (RequiredResourceAccess)
az ad app permission delete --id $appId --api 00000003-0000-0000-c000-000000000000
az ad app permission delete --id $appId --api 00000002-0000-0000-c000-000000000000

# Delete one --api-permissions (ResourceAccess)
az ad app permission delete --id $appId --api 00000003-0000-0000-c000-000000000000 --api-permissions 0e263e50-5827-48a4-b97c-d940288653c7
az ad app permission delete --id $appId --api 00000003-0000-0000-c000-000000000000 --api-permissions e1fe6dd8-ba31-4d61-89e7-88639da4683d
```

### CODE REVIEW

**Comment by yonzhan** [comment]
Role

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 1]
```diff
@@ -893,12 +893,32 @@ def add_permission(cmd, identifier, api, api_permissions):
...
+    # required_resource_access (list of RequiredResourceAccess)
+    #   RequiredResourceAccess
+    #     resource_app_id
+    #     resource_access (list of ResourceAccess)
+    #       ResourceAccess
+    #         id
+    #         type
+
+    # Get the RequiredResourceAccess object whose resource_app_id == api
+    access = next((a for a in required_resource_access if a.resource_app_id == api), None)
```
Comment (arrownj) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:909]: Can access be `None` ? If yes, we also need to add a check for it below.

Comment (jiasli) [reply src/azure-cli/azure/cli/command_modules/role/custom.py]: Nice catch. Added checks for both `api` and `api_permissions` and corresponding tests.

File: src/azure-cli/azure/cli/command_modules/role/tests/latest/test_graph.py [hunk 2]
```diff
@@ -616,4 +617,159 @@ def test_graph_required_access_e2e(self):
...
+            permissions = self.cmd('ad app permission list --id {app_id}').get_output_in_json()
+            self.assertEqual(permissions, [])
+
+            # Test delete non existing api permission from existing api
+            self.cmd('ad app permission add --id {app_id} --api {ms_graph_resource} '
+                     '--api-permissions {ms_target_api}=Scope {ms_target_api2}=Scope')
+            self.cmd('ad app permission delete --id {app_id} --api {ms_graph_resource} --api-permissions 22222222-0000-0000-c000-000000000000')
+            permissions = self.cmd('ad app permission list --id {app_id}').get_output_in_json()
+            ms_api = get_required_resource_access(permissions, self.kwargs['ms_graph_resource'])
+            self.assertEqual(ms_api['resourceAccess'], [ms_target_api_object, ms_target_api2_object])
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/tests/latest/test_graph.py:769]: > Can `access` be `None` ? If yes, we also need to add a check for it below.

Tests added.

File: src/azure-cli/azure/cli/command_modules/role/tests/latest/test_graph.py [hunk 3]
```diff
@@ -616,4 +617,159 @@ def test_graph_required_access_e2e(self):
...
+            permissions = self.cmd('ad app permission list --id {app_id}').get_output_in_json()
+            self.assertEqual(permissions, [])
+
+            # Test delete non-existing api permission from existing api
+            self.cmd('ad app permission add --id {app_id} --api {ms_graph_resource} '
+                     '--api-permissions {ms_target_api}=Scope {ms_target_api2}=Scope')
+            self.cmd('ad app permission delete --id {app_id} --api {ms_graph_resource} --api-permissions 22222222-0000-0000-c000-000000000000')
+            permissions = self.cmd('ad app permission list --id {app_id}').get_output_in_json()
+            ms_api = get_required_resource_access(permissions, self.kwargs['ms_graph_resource'])
+            self.assertEqual(ms_api['resourceAccess'], [ms_target_api_object, ms_target_api2_object])
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/tests/latest/test_graph.py:769]: > Can `access` be `None` ? If yes, we also need to add a check for it below.

Tests added.

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 4]
```diff
@@ -893,12 +893,47 @@ def add_permission(cmd, identifier, api, api_permissions):
...
+    #         id
+    #         type
+
+    # Get the RequiredResourceAccess object whose resource_app_id == api
+    rra = next((a for a in required_resource_access if a.resource_app_id == api), None)
+
+    if not rra:
+        # Silently pass if the api is not required.
+        logger.warning("App %s doesn't require access to API %s.", identifier, api)
+        return
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:914]: Check the presence of `api`.

File: src/azure-cli/azure/cli/command_modules/role/custom.py [hunk 5]
```diff
@@ -893,12 +893,47 @@ def add_permission(cmd, identifier, api, api_permissions):
...
+    if api_permissions:
+        # Check if the user tries to delete any ResourceAccess that is not required.
+        ra_ids = [ra.id for ra in rra.resource_access]
+        non_existing_ra_ids = [p for p in api_permissions if p not in ra_ids]
+        if non_existing_ra_ids:
+            logger.warning("App %s doesn't require access to API %s's permission %s.",
+                           identifier, api, ', '.join(non_existing_ra_ids))
+            if len(non_existing_ra_ids) == len(api_permissions):
+                # Skip the REST call if nothing to remove
+                return
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/custom.py:925]: Check the presence of `api-permissions`.

**Review by arrownj — APPROVED** [review]
💯

**Review by jsntcy — APPROVED** [review]
:shipit:

**Pull request merged**